server/plugin.ts

```diff
@@ -138,6 +144,12 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
 
     // set up multi-tenant routes
     if (config.multitenancy?.enabled) {
+      if (workspace) {
+        this.logger.error(
+          'Both workspace and multi-tenancy features are enabled, only one of them can be enabled at the same time.'
+        );
+        process.exit(1);
```

Comment:
Why are we killing OSD? Is there a better way here? Can we default to only setup workspace and log this out instead of killing the process?

Reply:
> Why are we killing OSD? Is there a better way here?

The idea is that if we don't kill the process, user still use multi tenancy and workspace at the same time if they choose to ignore the error or warning message.


> Can we default to only setup workspace and log this out

does this mean we don't setup multi tenancy if workspace is enabled?

Reply:
Yea maybe it is a product/UX question but from my perspective killing the process seems bad. If we are encouraging people to migrate to workspaces anyway, why can't we default to workspaces and simply tell them we are doing so? Since workspaces is a new feature and opt in I don't think this is a big stretch - @kgcreative have any thoughts here?

Reply:
Most likely user doesn't know workspace and multi tenant are mutually exclusive if they have both feature flags on. If we turn off multi tenant automatically silently, will it be a problem? I feel it's better to stop the booting process to let user make the decision.  User should already stop the OSD process before they make the configuration change and just restart the process so the impact to stop it is not serious.

Reply:
We should add a toggle here instead of killing the process. When we enable one we automatically disable the other one and log that out.

Dynamic tenancy configurations #2607 - this PR allows enabling / disabling multi-tenancy dynamically.

Reply:
I'm ok with killing the process on startup, but its not enough to only check the configuration in `opensearch_dashboards.yml`. You can think of that as a feature flag for dynamic multi-tenancy, rather than the actual value if tenancy is enabled or disabled. If the feature flag is false, then it definitely means that multitenancy is disabled, but if its true then you need to obtain the actual value from the backend.

There's an endpoint called `/dashboardsinfo` to get dynamic security settings and you can see an example of calling it [here](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/plugin.ts#L94).

Reply:
@Hailong-am wdyt about disabling the ability to toggle multi-tenancy dynamically from off -> on if workspaces is enabled?

The code for dynamic multi-tenancy UX can be found here: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/configuration/panels/tenant-list/save_changes_modal.tsx#L84-L88

Reply:
> @Hailong-am wdyt about disabling the ability to toggle multi-tenancy dynamically from off -> on if workspaces is enabled?
> 
> The code for dynamic multi-tenancy UX can be found here: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/configuration/panels/tenant-list/save_changes_modal.tsx#L84-L88

Yeah, make sense.

Reply:
@cwperks @DarshitChanpura I have update the PR upon your suggestions, could you help to review it again?

Reply:
>for a service to handle an invalid configuration after it has started.

@peternied  For invalid configuration, if we have unrecognized configuration in yml file OSD will crash when start up. That's current behavior. The check we added here is slightly different, but we can follow the same pattern.

An example
```
  log   [05:27:15.151] [fatal][root] InvalidConfigurationError: Unknown configuration key(s): "abc.enabled". Check for spelling errors and ensure that expected plugins are installed.
    at ensureValidConfiguration (/Users/ihailong/code/OpenSearch-Dashboards-workspace/src/core/server/legacy/config/ensure_valid_configuration.ts:50:11)
    at Server.setup (/Users/ihailong/code/OpenSearch-Dashboards-workspace/src/core/server/server.ts:152:5)
    at Root.setup (/Users/ihailong/code/OpenSearch-Dashboards-workspace/src/core/server/root/index.ts:64:14)
    at bootstrap (/Users/ihailong/code/OpenSearch-Dashboards-workspace/src/core/server/bootstrap.ts:132:5)
    at Command.<anonymous> (/Users/ihailong/code/OpenSearch-Dashboards-workspace/src/cli/serve/serve.js:279:5) {
  code: 'InvalidConfig',
  processExitCode: 64,
  cause: undefined
}

 FATAL  Error: Unknown configuration key(s): "abc.enabled". Check for spelling errors and ensure that expected plugins are installed.

```

Here are the options we have think about:
1. Document the prerequisites of enabling workspace in document site and let user to do the right configuration
2. Instead of exit the process, log the warning message. Most of time cx will choose to ignore warning, even for today when OpenSearch dashboard start there have few warnings, not everyone cares about warning message.
3. Disable multi-tenancy automatically when workspace is enabled

Option 1 and 2 will not enforce cx to just one of them and have the possibility to mix them together, we don't want to support this mixed situation.

Option 3, As @xluo-aws mentioned here https://github.com/opensearch-project/security-dashboards-plugin/pull/1822#discussion_r1523334317 , we don't know the intention of cx and disable multi-tenancy by default is not a good decision.